Align searcher with name/expr. Refactor CSS, remove magic constants.

piotrMocz opened the pull request:
Okay, this rights some of the wrongs that happened to the searcher styles (and were mostly a result of transfering to the new gui quickly and patching the existing styles.

The changes this introduces can be summarised as follows:
* When you double click on a node's name or expression, the searcher input is exactly in the same place.
* `basegl-ui` does not set the position/offset of the searcher in any way! The css/html takes care of positioning the base of the searcher component exactly in the center of the `input` field. Basegl assumes that the `(0, 0)` of a component is its geometrical center, so that aligns well.
* The previous allows for removal of some bad-looking hardcoded offsets from the `basegl-ui`.
* The CSS itself is arguably cleaner and more concise: we no longer rely on the `transform` property and absolute positioning, setting the position to relative with respect to one common container.

I'm quite proud of these changes as this kind of styling work is pretty far away from my comfort zone 😂

ljkania commented on dist/style/_searcher.less:
```diff
@@ -7,45 +7,62 @@
 
-@searcherWidth:      400px;
+@entryHeight:    @luna-superline-height * 1px; // calc(@ui ~'*' @luna-superline-height * 1px);
+@borderRadius:   (@luna-searcher-border-radius + @luna-connection-width) * 1px;
```
`* 1px` - what is this magic? :O 
If this is just for adding `px` to the number then probably there is a better way to express it. If `@luna-superline-height` is number then `@entryHeight` should be as well. Also, there is some inconsistency in the naming convention - why we use camel case and `-` the same time?

piotrMocz replied:
Ok, so "height" should be a value with a unit. `12` is not a height, `12px` is. I don't want to touch too much of existing code, so I'm adopting the camelCase convention, like the whole codebase.

ljkania replied:
Yeah, the only thing is that I don't like multiplication by `1px`.
See this: https://stackoverflow.com/questions/8736757/how-can-i-have-multiply-a-pixel-width-value-with-less-css/25774297 - as far as I understand the best answer, if you have a raw value assigned to a variable somewhere (like `@searcherWidth: 400px`), all the later operations should preserve `px`. Which makes sense considering that you achieve the same by multiplying it `1px` times.
Also see this: https://stackoverflow.com/questions/16625060/how-to-get-value-in-percentage-or-pixel-px-after-doing-some-maths-lessc - best answer here suggest a nice way of denoting that you are specifying units. I like it much more better than `*1px` but I accept if you feel different.
And one more thing: Indentation here is too big. Please make it one space after the longest name or drop indentation here completely - it is not very needed here.

ljkania commented on dist/style/_searcher.less:
```diff
@@ -7,45 +7,62 @@
+// 1. move the whole searcher down by @searcherHeight
+// 2. move it up by 1/2 of the height of the input field.
+// Now we are in the middle of the input field and everything is aligned.
+@searcherCenterOffset: -10 * @entryHeight / 2;
```
I love this comment, really! It is nice and accurate and it makes a lot of sense! I understand where this comes from! <3
What is more, this helps me spot this problem: why `@searcherCenterOffset` is not simply `-@searcherHeight + @entryHeight/2`? Also, the math does not add up, check this:
`-@searcherHeight + @entryHeight/2 = -11 * @entryHeight + @entryHeight/2 = -21/2 * @entryHeight`
It is certainly different than `-10 * @entryHeight / 2` you got here.

ljkania replied:
Yeah, I realize this is correct now. However totally wrong to use different values than in description. We don't lose much of the performance doing simply: `-@searcherHeight / 2 + @entryHeight / 2` but if you worried about division performance you can do this `(-@searcherHeight - @entryHeight) / 2` - add is almost for free.
We shouldn't count it like you do, because it is not what our logic says. If we ever decided to change `@searcherHeight` we will need to change this line again, because logic will be the same. Let's keep our code self-maintainable wherever possible.

ljkania commented on dist/style/_searcher.less:
```diff
@@ -7,45 +7,62 @@
+    width:  @searcherWidth;
+    height: @searcherHeight;
+    position: relative;
+    transform: translateY(@searcherCenterOffset);
```
Indent or not, just be concise in whole file ;)

ljkania commented on dist/style/_searcher.less:
```diff
@@ -55,35 +72,36 @@
+    // NOTE: there used to be a (transparent) border around the results.
+    // however, it adds padding between the results and the input,
+    // which in my opinion doesn't look well. If you want it back, uncomment:
+    // border: unit(@luna-connection-width, px) solid transparent;
```
Check this with @wdanilo. Let's get things done at earliest convenience which happens now.

ljkania commented on dist/style/_searcher.less:
```diff
@@ -55,35 +72,36 @@
-    width:          calc(@ui ~'*' @searcherWidth);
-    max-height:     calc(@ui ~'*' @luna-superline-height * 10px);  
+    width:          @searcherWidth;
+    max-height:     10 * @entryHeight;  
```
I love this change <3 NO MORE `calc`

ljkania commented on src/shape/Html.coffee:
```diff
@@ -10,6 +10,7 @@ export class HtmlShape extends BasicComponent
         scalable: true
         still: false
         clickable: true
+        cssClassName: []
```
Clean, neat, I like it.

pmlodawski commented on src/shape/Html.coffee:
```diff
@@ -18,6 +19,7 @@ export class HtmlShape extends BasicComponent
         root = document.createElement @model.element
         root.id = @model.id if @model.id?
         root.style.pointerEvents = if @model.clickable then 'all' else 'none'
+        root.className = @model.cssClassName
```
This will update only once, since `redefineRequired` does not say that it should run `define` again, if `@changed.cssClassName`. I fixed HtmlShape logic here, so you can apply this fix here too. https://github.com/luna/basegl-ui/pull/28

pmlodawski commented on src/view/ExpressionNode.coffee:
```diff
@@ -203,8 +203,8 @@ export class ExpressionNode extends ContainerComponent
     makeHoverable: (view) =>
-        view.addEventListener 'mouseenter', => @set hovered: true
-        view.addEventListener 'mouseleave', => @set hovered: false
+        view.addEventListener 'mouseenter', => @set(hovered: true)
```
unneccessary braces

pmlodawski commented on src/view/EditableText.coffee:
```diff
@@ -50,11 +46,11 @@ export class EditableText extends ContainerComponent
 
     hideSearcher: =>
-        @set edited: false
+        @set(edited: false)
```
unnecessary braces

pmlodawski commented on src/view/EditableText.coffee:
```diff
@@ -50,11 +46,11 @@ export class EditableText extends ContainerComponent
 
     showSearcher: (notify = true) =>
-        @set edited: true
+        @set(edited: true)
```
unnecessary braces

pmlodawski commented on src/view/ExpressionNode.coffee:
```diff
@@ -203,8 +203,8 @@ export class ExpressionNode extends ContainerComponent
-        view.addEventListener 'mouseenter', => @set hovered: true
-        view.addEventListener 'mouseleave', => @set hovered: false
+        view.addEventListener 'mouseenter', => @set(hovered: true)
+        view.addEventListener 'mouseleave', => @set(hovered: false)
```
unnecessary braces

piotrMocz replied:
But the linter complains and colors my files red :(
```
[coffee] 239:22 'set' cannot be used as a keyword, or as a function call without parentheses
```

piotrMocz commented:
I implemented @ljkania 's suggestions, removed all of the`calc`-s from the code and cleaned it up. Also, I finally caved and converted all of the LESS variable names to `this-notation` to be consistent with the rest of the LESS files. FWIW this is called `kebab-case` and I love the name :D 🌮
